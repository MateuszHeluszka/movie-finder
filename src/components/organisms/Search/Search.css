.search {
  width: 100%;
  padding: 40px 5%;
}
