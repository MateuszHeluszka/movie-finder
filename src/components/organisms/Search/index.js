import React, { Component } from 'react';
import { findMovies, getMovie } from '../../../services/api';
import SearchBar from '../../molecules/SearchBar';
import SearchResults from '../../molecules/SearchResults';
import './Search.css';

class Search extends Component {
  state = {
    value: '',
    error: false,
    loading: false,
    debounce: null,
    message: '',
    results: [],
    details: {
      visible: false,
      data: null,
    },
  };

  onChange = e => {
    /**
     * Handle input change
     *
     * @param {function} e
     * @public
     */

    if (this.state.debounce) {
      clearTimeout(this.state.debounce);
    }

    this.setState({
      value: e.target.value,
      message: '',
      debounce: setTimeout(() => this.search(this.state.value), 1000),
    });
  };

  search = query => {
    /**
     * Call api to find movies by passed query
     *
     * @param {string} query
     * @public
     */

    if (query !== undefined && query !== '') {
      this.setState({
        loading: true,
      });

      findMovies(`${query}`)
        .then(res => {
          if (res.total_results === 0) {
            this.setState({
              loading: false,
              error: false,
              results: [],
              message: 'No match results.',
            });
          } else {
            this.setState({
              loading: false,
              results: [...res.results],
            });
          }
        })
        .catch(() => {
          this.setState({
            error: true,
            loading: false,
          });
        });
    } else {
      this.setState({
        results: [],
        error: false,
      });
    }
  };

  showDetails = id => {
    /**
     * Call api to find movie details and show modal
     *
     * @param {number} id
     * @public
     */
    getMovie(id)
      .then(res => {
        this.setState({
          details: {
            visible: true,
            data: res,
          },
        });
      })
      .catch(error => {
        this.setState({
          message: error,
        });
      });
  };

  hideDetails = () => {
    this.setState({
      details: {
        visible: false,
        data: null,
      },
    });
  };

  render() {
    return (
      <section className="search">
        <SearchBar
          loading={this.state.loading}
          onChange={this.onChange}
          value={this.state.value}
          type="search"
        />
        {this.state.error === true && (
          <p className="error">An error occured, try again.</p>
        )}
        <SearchResults
          showDetails={this.showDetails}
          hideDetails={this.hideDetails}
          details={this.state.details}
          items={this.state.results}
          message={this.state.message}
        />
      </section>
    );
  }
}

export default Search;
