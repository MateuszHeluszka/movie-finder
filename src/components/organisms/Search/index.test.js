import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Search from './index';

test('Component renders corectly', () => {
  const component = renderer.create(
    <Search />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
