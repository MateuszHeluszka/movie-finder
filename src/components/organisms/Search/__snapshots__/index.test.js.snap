// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<section
  className="Search"
>
  <section
    className="SearchBar"
  >
    <input
      onChange={[Function]}
      type="search"
      value=""
    />
  </section>
</section>
`;
